#######
# This script computes raw differences in genotypes between duplicates in the UK Biobank.
#######

basedir=/well/ukbiobank/qcoutput.V2_QCed.sample-QC
plink=/well/ukbiobank/qcoutput/Src/plink-v1.90b3k-x86_64/plink

# The genotype data to use
#genoData=$basedir/data/Combined/b1__b11-b001__b095-autosome-sampleqc
#genoData=/well/ukbiobank/expt/V2_QCed.export/data/genotype_gwas/V2_QCed.export.UKBiLEVEAX_b1-b11.Batch_b001_b095.genotype_gwas
genoData=/well/ukbiobank/expt/V2_QCed.export/data/calls_export/v2/V2_QCed.export.UKBiLEVEAX_b1-b11.Batch_b001_b095.no_sample_excl  # <======= CONFIRMED WITH COLIN. NOTE: will not have excluded extra array-fail X chromosome SNPs. v3 is the real released set, but the only difference is the X chrom. snps.

# Which version of the king output are we using?
#kingRunPrefix=b1__b11-b001__b095-pair_batches.filtered-samples-pruned-200
kingRunPrefix=b1__b11-b001__b095-pair_batches.filtered

# All the twins and duplicates (file generated by find-families.R)
duplicates=$basedir/QC-Scripts/Sample-QC/Relatedness/${kingRunPrefix}-duplicates-twins.txt


# Just compute distances for all of them and prune out genuine twins later.
#cat <(tail -n +2 $duplicates | head -n $dup | tail -n 1 | cut -f2 -d' ') <(tail -n +2 $duplicates | head -n $dup | tail -n 1 | cut -f1 -d' ') | awk '{print $0,$0}' > $basedir/QC-Scripts/Sample-QC/Relatedness/keep.$dup.txt

#cat <(grep -w `tail -n +2 $duplicates | head -n $dup | tail -n 1 | cut -f1 -d' '` $genoData.fam | cut -f1,2 -d' ') <(grep -w `tail -n +2 $duplicates | head -n $dup | tail -n 1 | cut -f2 -d' '` $genoData.fam | cut -f1,2 -d' ') > $basedir/QC-Scripts/Sample-QC/Relatedness/keep.$dup.txt

#grep -wE -f <( tail -n +2 $duplicates | cut -f1,2 -d' ' | tr ' ' '\n') $genoData.fam | cut -f1,2 -d' ' > $basedir/QC-Scripts/Sample-QC/Relatedness/keep.duplicates.txt

grep -wE -f <( tail -n +2 $duplicates | cut -f2,4 -d' ' | tr ' ' '\n') $genoData.fam | cut -f1,2 -d' ' > $basedir/QC-Scripts/Sample-QC/Relatedness/keep.duplicates.txt

# 1715 unique duplicated samples

# NOTE: Only compute on markers where both individuals are non-missing. Takes a couple of minutes to run for two individuals! Also only compute on the autosomes!

# $plink --bfile $genoData --keep $basedir/QC-Scripts/Sample-QC/Relatedness/keep.$dup.txt --chr 1-22 --geno 0 --distance --out $basedir/data/Relatedness/${kingRunPrefix}-duplicates-genetic-distances-release-$dup

#$plink --bfile $genoData --keep $basedir/QC-Scripts/Sample-QC/Relatedness/keep.duplicates.txt --chr 1-22 --distance square0 ibs allele-ct flat-missing --out $basedir/data/Relatedness/${kingRunPrefix}-duplicates-genetic-distances-release


$plink --bfile $genoData --keep $basedir/QC-Scripts/Sample-QC/Relatedness/keep.duplicates.txt --chr 1-22 --genome full --out $basedir/data/Relatedness/${kingRunPrefix}-duplicates-genetic-distances-release


# This is a large file with lots of extraneous information. Extract duplicate pairs only.
Rscript ./postking/subsetPlinkIBS.R $basedir/data/Relatedness/${kingRunPrefix}-duplicates-genetic-distances-release.genome $duplicates

rm $basedir/data/Relatedness/${kingRunPrefix}-duplicates-genetic-distances-release.genome


#rm $basedir/data/Relatedness/${kingRunPrefix}-duplicates-genetic-distances-release-$dup.dist.id


# Write out R-readible file
## $plink --bfile $genoData --keep keep.$dup.txt --recode A-transpose --out test-$dup


# For duplicate set 1:

# diff=620
# 40583 variants removed due to missing genotype data (--geno).
# 743673 variants and 2 people pass filters and QC.


#### CLARE ADDED BELOW IN 2018 (only ran below)

# Write out R-readible file
$plink --bfile $genoData --keep $basedir/QC-Scripts/Sample-QC/Relatedness/keep.duplicates.txt --recode A-transpose --out $basedir/data/Relatedness/${kingRunPrefix}-duplicates-genotypes


Rscript ./postking/subsetPlinkIBS_bySNP.R $basedir/data/Relatedness/${kingRunPrefix}-duplicates-genotypes.traw $duplicates

# NOW submit difference calculation to the cluster! e.g 

# qsub -t 1-$nDupes -P donnelly.prjc -q short.qc -N dup-concordance-bySNP
